src/server/managers/AppLicenseManager.ts

```diff
@@ -0,0 +1,79 @@
+import { AppManager } from '../AppManager';
+import { IUserBridge } from '../bridges';
+import { InvalidLicenseError } from '../errors';
+import { IMarketplaceInfo } from '../marketplace';
+import { AppLicenseValidationResult } from '../marketplace/license';
+import { Crypto } from '../marketplace/license';
+
+enum LicenseVersion {
+    v1 = 1,
+}
+
+export class AppLicenseManager {
+    private readonly crypto: Crypto;
+    private readonly userBridge: IUserBridge;
+    constructor(private readonly manager: AppManager) {
+        this.crypto = new Crypto(this.manager.getBridges().getInternalBridge());
+        this.userBridge = this.manager.getBridges().getUserBridge();
+    }
+
+    public async validate(validationResult: AppLicenseValidationResult, appMarketplaceInfo?: IMarketplaceInfo): Promise<void> {
+        if (!appMarketplaceInfo || !appMarketplaceInfo.subscriptionInfo) {
+            return;
+        }
+
+        validationResult.setValidated(true);
+
+        const { id: appId, subscriptionInfo } = appMarketplaceInfo;
+
+        let license;
+        try {
+            license = await this.crypto.decryptLicense(subscriptionInfo.license.license) as any;
+        } catch (err) {
+            validationResult.addError('publicKey', err.message);
+
+            throw new InvalidLicenseError(validationResult);
+        }
+
+        if (license.appId !== appId) {
+            validationResult.addError('appId', `License hasn't been issued for this app`);
```

Comment:
I wonder about using the context phrase "this" when using error codes. Guess it depends on how the i18n will work for these and how the errors/warnings will be displayed.

Reply:
Great point, certainly haven't taken that into account. Maybe the engine return error codes instead of messages, and then Rocket.Chat can apply i18n?

Reply:
Possibly? Maybe return both? Better to be verbose than not when it comes to trying to determine where a bug is..

Reply:
Well, then I guess it's ok - Rocket.Chat will be able to differentiate the errors with the `field` property (`'appId'` in this case)